From d8d343374ba1440fb9fe011566f0663cd8f7a43a Mon Sep 17 00:00:00 2001
From: Yan Wang <yan.wang@linux.intel.com>
Date: Wed, 21 Mar 2018 17:01:34 +0800
Subject: [PATCH 0870/1240] ASoC: sof: Fix oops due to trace_sleep use before
 init

Access happens on the following path:

1) sof_ipc_tx_message
  => tx_wait_done
    => snd_sof_trace_notify_for_error
     => wake_up(&sdev->trace_sleep)

where sdev->trace_sleep is not yet initialized.

While at it also move initialization of sdev->host_offset
before starting working with sdev struct.

Signed-off-by: Daniel Baluta <daniel.baluta@gmail.com>
---
 sound/soc/sof/trace.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/sound/soc/sof/trace.c b/sound/soc/sof/trace.c
index ac50a75..5bb12b7 100644
--- a/sound/soc/sof/trace.c
+++ b/sound/soc/sof/trace.c
@@ -203,6 +203,9 @@ int snd_sof_init_trace(struct snd_sof_dev *sdev)
 	params.buffer.offset = 0;
 	params.buffer.pages = sdev->dma_trace_pages;
 
+	init_waitqueue_head(&sdev->trace_sleep);
+	sdev->host_offset = 0;
+
 	/* send IPC to the DSP */
 	ret = sof_ipc_tx_message(sdev->ipc,
 				 params.hdr.cmd, &params, sizeof(params),
@@ -213,8 +216,6 @@ int snd_sof_init_trace(struct snd_sof_dev *sdev)
 		goto table_err;
 	}
 
-	init_waitqueue_head(&sdev->trace_sleep);
-	sdev->host_offset = 0;
 	sdev->dtrace_is_enabled = true;
 	return 0;
 
-- 
2.7.4

